### DISCUSSION TITLE
Sample to load model and run inference?

### DISCUSSION THREAD

**Discussion by eliben** [gomlx/gomlx#86] [Q&A] [answered]
I was able to follow the tutorial and piecing together code from `example/cifar` to train a CNN for CIFAR-10 (I have a local setup for this that doesn't use notebooks/jupyter).

However, it's not clear to me how to write an inference tool that loads a model from a checkpoint and can run inference on images using it; is there a sample for this, or documentation?

**Comment by janpfeifer** [comment]
It's good to hear that the tutorial went fine. About saving the model:

**Current**: the model is the code. Saving a model means loading the Go library with the code that builds the model, load the checkpoint, JIT-compiling it and using it. This is the most generic form of saving models, because it allows a lot to be dynamically (on-the-fly) defined according to inputs. But maintaining versioning of code hand-in-hand with checkpoints require lots of dicipline.

**Future**: definitely we want exchange formats that don't require sending code. The first 2 in consideration are:

* (A) HLO (or StableHLO) + checkpoint: the code already generates that, so it should be easy. Inference will still need to link the PJRT plugin (at least with the XLA engine). Also a down-side is that the HLO has to be saved to a fixed shape (or a fixed number of shapes). It's probably fine for the majority of cases, but not every case.
*  (B) ONNX: so other runtime may be used to serve it, [github.com/gomlx/onnx-gomlx](https://github.com/gomlx/onnx-gomlx) has gone a long way converting ONNX models to GoMLX, but the other way around still requires lots of work.

Let me know if you have a immediate use-case for having some saved model format, and if format (A) HLO proto buf+checkpoint would suffice. I could put this together in a week, and I would prioritize that if there is a motivating case 😃

**Comment by eliben** [comment]
Thanks for the quick reply; I'm OK with the current state - is there a sample for the "load the checkpoint then JIT-compile and use it" part?

**Comment by janpfeifer** [comment, marked as answer]
I just created a demo for it -- it will come in the next release, but it's already in HEAD:

* A [classifier package for the Cifar demo](https://github.com/gomlx/gomlx/blob/main/examples/cifar/classifier/classifier.go), which would be the typical usage case.
* Calling the inference code [in the notebook](https://gomlx.github.io/gomlx/notebooks/cifar.html) -- see "Inference" section, in cell 13 close to the very bottom.

Let me know if you have any questions or suggestion 😃 

ps.: I created a new `checkpoints.Load()` that is identical to `checkpoints.Build`, except it complains if the checkpoint doesn't exist. It will make to the next release (in a few days) or one can get it from head.

**Reply by eliben** [reply]
Thanks for creating the classifier demo!

Trying to reproduce it (outside a notebook), I had to comment out this line:

```
c.ctx = c.ctx.Reuse()
```

Otherwise I get an error:

> requested variable "weights" in scope "/model/000_conv/conv" with Context.Reuse set, but variable does not exist

Now it runs, but doesn't appear to classify properly (it returns random-ish classes, different ones for the same image if I rerun classification multiple times). I have to debug this to see what's different from your notebook version...

**Reply by janpfeifer** [reply]
Ops, I should have mentioned something, the Cifar example had a minor bug where I was creating the variables under the scope `/model/model`. So I changed to create the variables under `/model`.

I assume your model was created before my recent update ? So the error was correct: it was not finding the variable. 

Three options here:

1. Change the `classifier` code where it read `ctx.In("model")` replace with `ctx.In("model").In("model")`. Alternatively, inspect the model and decide on-the-fly which one to use.
2. The variable names are stored in a .json file in the checkpoints. You may as well just open it in an editor and do a search&replace.
3. Retrain the model with the new code, it will create the variables in the right place.

There is a command line tool called `gomlx_checkpoints` under `cmd/` directory, that you can use to list print a summary of the model, list the variables (with shape and range of values), etc. Check it out. To install: `go install github.com/gomlx/gomlx/cmd/gomlx_checkpoints@latest`. And to list variables: `gomlx_checkpoints --vars <checkpoint_dir>`. Use `--help` for help.

I would keep the `c.ctx.Reuse()` line, it's there exactly to point out these problems. What happened in your case is that, since the variables didn't exits, it was creating the variables with random initialization. You can check that because each time your run it you'll get a different prediction.

**Reply by eliben** [reply]
Thanks, I managed to get it to work now with your suggestion.

I'm trying to clean up and simplify the demo; considering to write a blog post about it.

**Reply by janpfeifer** [reply]
Nice, I curious to see it. Let me know if you need some help.

**Reply by janpfeifer** [reply]
Just to wrap up the topic, the inference example, along with a section in the `README.md` file included in the v0.15.2 release.

**Reply by eliben** [reply]
> Nice, I curious to see it. Let me know if you need some help.

Here it is: https://eli.thegreenplace.net/2024/gomlx-ml-in-go-without-python/

**Reply by janpfeifer** [reply]
It is great Eli! Very nice introduction of the layers of an ML system, and the examples are perfect: one image with Cifar and I'm so happy you added the Gemma2 model as well.

This is really awesome!
